packages/near-membrane-base/src/blue.ts

```diff
@@ -161,52 +154,99 @@ export function blueProxyFactory(env: MembraneBroker) {
         // istanbul ignore else
         if ('writable' in blueDescriptor) {
             // We are dealing with a value descriptor.
-            const { value } = blueDescriptor;
-            blueDescriptor.value =
-                typeof value === 'function'
-                    ? // We are dealing with a method (optimization).
-                      getBlueFunction(value)
-                    : getBlueValue(value);
+            blueDescriptor.value = getBlueValue(blueDescriptor.value);
         } else {
             // See additional details above.
             // We are dealing with accessors.
             const { get, set } = blueDescriptor;
-            if (typeof get === 'function') {
-                blueDescriptor.get = getBlueFunction(get);
+            if (get !== undefined) {
+                blueDescriptor.get = getBlueValue(get);
             }
-            if (typeof set === 'function') {
-                blueDescriptor.set = getBlueFunction(set);
+            if (set !== undefined) {
+                blueDescriptor.set = getBlueValue(set);
             }
         }
         return blueDescriptor;
     }
 
     function getBlueFunction(redFn: RedFunction): BlueFunction {
-        const blueFn = env.getBlueRef(redFn);
-        if (blueFn === undefined) {
-            return createBlueProxy(redFn) as BlueFunction;
+        // caching logic
+        const blueFn = WeakMapGet(env.redMap, redFn) as RedFunction | undefined;
+        if (blueFn !== undefined) {
+            return blueFn;
+        }
+        // extracting the metadata about the proxy target
+        const targetTypeof = 'function';
+        let targetIsArrowFunction = false;
+        let targetFunctionName: string | undefined;
+        const targetIsArray = false;
+        // detecting arrow function vs function
+        try {
+            targetIsArrowFunction = !('prototype' in redFn);
+        } catch {
+            // target is either a revoked proxy, or a proxy that throws on the
+            // `has` trap, in which case going with a strict mode function seems
+            // appropriate.
         }
-        return blueFn as RedFunction;
+        try {
+            // a revoked proxy will throw when reading the function name
+            targetFunctionName = ReflectGetOwnPropertyDescriptor(redFn, 'name')?.value;
+        } catch {
+            // intentionally swallowing the error because this method is just extracting the
+            // function in a way that it should always succeed except for the cases in which
+            // the provider is a proxy that is either revoked or has some logic to prevent
+            // reading the name property descriptor.
+        }
+        return createBlueProxy(
+            redFn,
+            targetTypeof,
+            targetIsArrowFunction,
+            targetFunctionName,
+            targetIsArray
+        ) as BlueFunction;
+    }
+
+    function getBlueObjectOrArray(red: RedArrayOrObject): BlueArrayOrObject {
```

Comment:
Since the returned type is called `BlueArrayOrObject`, lets call the function `getBlueArrayOrObject`